File: SOURCE_LEAN/EW4_Web_Port_v0.61/native_conquest_extinction_core.js
'use strict';
(function(root,factory){
  const api=factory();
  if(typeof module==='object'&&module.exports)module.exports=api;
  root.EW4NativeConquestExtinction=api;
})(typeof globalThis!=='undefined'?globalThis:this,function(){
  const LAND_ARMY_TYPES=new Set(['infantry','cavalry','artillery']);
  const LAND_FACILITY_TYPES=new Set(['city','industry','stable','farmland']);
  const PORT_TYPE='port';
  function ownerOf(x){return x?.owner==null?255:+x.owner}
  function alive(u){return !!u&&!u.dead&&(+u.hp||0)>0}
  function typeOf(u,unitType){const t=typeof unitType==='function'?unitType(u):u?.type;return String(t||'').toLowerCase()}
  function isLandArmy(u,unitType){return alive(u)&&LAND_ARMY_TYPES.has(typeOf(u,unitType))}
  function isCountedFacility(o){return !!o&&!!o.construction_type&&(LAND_FACILITY_TYPES.has(String(o.construction_type))||String(o.construction_type)===PORT_TYPE)}
  function countryStatus(state,owner,unitType){
    owner=+owner;
    const units=state?.units||[],objects=state?.objects||[];
    const landArmyAlive=units.some(u=>ownerOf(u)===owner&&isLandArmy(u,unitType));
    const landFacilitiesHeld=objects.some(o=>ownerOf(o)===owner&&LAND_FACILITY_TYPES.has(String(o?.construction_type||'')));
    const portsHeld=objects.some(o=>ownerOf(o)===owner&&String(o?.construction_type||'')===PORT_TYPE);
    return{owner,landArmyAlive,landFacilitiesHeld,portsHeld,defeated:!landArmyAlive&&!landFacilitiesHeld&&!portsHeld};
  }
  function ownerCandidates(state){
    const out=new Set();
    for(const c of state?.battle?.countries||[])if(c?.index!=null&&+c.index!==255)out.add(+c.index);
    for(const u of state?.units||[])if(ownerOf(u)!==255)out.add(ownerOf(u));
    for(const o of state?.objects||[])if(isCountedFacility(o)&&ownerOf(o)!==255)out.add(ownerOf(o));
    return [...out].sort((a,b)=>a-b);
  }
  function defeatedOwners(state,unitType){return ownerCandidates(state).filter(o=>countryStatus(state,o,unitType).defeated)}
  function hostileOwners(state,playerOwner,relation){
    return ownerCandidates(state).filter(owner=>owner!==+playerOwner&&typeof relation==='function'&&relation(+playerOwner,owner)==='hostile');
  }
  function outcome(state,{playerOwner=0,unitType,relation}={}){
    if(!state)return null;
    const player=countryStatus(state,+playerOwner,unitType);
    if(player.defeated)return{kind:'defeat',reason:'country-extinction',owner:+playerOwner,status:player};
    const hostile=hostileOwners(state,+playerOwner,relation);
    if(hostile.length&&hostile.every(o=>countryStatus(state,o,unitType).defeated))return{kind:'victory',reason:'hostile-powers-extinct',owners:hostile};
    return null;
  }
  return{LAND_ARMY_TYPES,LAND_FACILITY_TYPES,PORT_TYPE,isLandArmy,isCountedFacility,countryStatus,ownerCandidates,defeatedOwners,hostileOwners,outcome};
});

File: SOURCE_NATIVE/EW4_iOS_NativePort/NativeCore/Sources/EW4NativeCore/NativeConquestExtinctionCore.swift
import Foundation

public struct NativeConquestCountryStatus: Equatable, Sendable {
    public let owner: Int
    public let landArmyAlive: Bool
    public let landFacilitiesHeld: Bool
    public let portsHeld: Bool
    public var defeated: Bool { !landArmyAlive && !landFacilitiesHeld && !portsHeld }
}

public enum NativeConquestOutcome: Equatable, Sendable {
    case victory(reason: String, owners: [Int])
    case defeat(reason: String, owner: Int, status: NativeConquestCountryStatus)
}

public enum NativeConquestExtinctionCore {
    public static let landArmyTypes: Set<String> = ["infantry", "cavalry", "artillery"]
    public static let landFacilityTypes: Set<String> = ["city", "industry", "stable", "farmland"]
    public static let portType = "port"

    public static func countryStatus(_ state: NativeBattleGameplayState, owner: Int) -> NativeConquestCountryStatus {
        let landArmyAlive = state.unitOrder.compactMap { state.units[$0] }.contains { unit in
            guard unit.owner == owner, !unit.dead, let stat = state.stat(for: unit) else { return false }
            return landArmyTypes.contains(stat.type.lowercased())
        }
        let landFacilitiesHeld = state.objects.values.contains { object in
            object.owner == owner && landFacilityTypes.contains((object.constructionType ?? "").lowercased())
        }
        let portsHeld = state.objects.values.contains { object in
            object.owner == owner && (object.constructionType ?? "").lowercased() == portType
        }
        return NativeConquestCountryStatus(
            owner: owner,
            landArmyAlive: landArmyAlive,
            landFacilitiesHeld: landFacilitiesHeld,
            portsHeld: portsHeld
        )
    }

    public static func ownerCandidates(_ state: NativeBattleGameplayState) -> [Int] {
        var owners = Set<Int>()
        for country in state.battle.countries where country.index != 255 { owners.insert(country.index) }
        for unit in state.units.values where unit.owner != 255 { owners.insert(unit.owner) }
        for object in state.objects.values {
            let type = (object.constructionType ?? "").lowercased()
            if object.owner != 255 && (landFacilityTypes.contains(type) || type == portType) {
                owners.insert(object.owner)
            }
        }
        return owners.sorted()
    }

    public static func defeatedOwners(_ state: NativeBattleGameplayState) -> [Int] {
        ownerCandidates(state).filter { countryStatus(state, owner: $0).defeated }
    }

    public static func hostileOwners(_ state: NativeBattleGameplayState) -> [Int] {
        ownerCandidates(state).filter { owner in
            owner != state.playerOwner && state.relation(state.playerOwner, owner) == .hostile
        }
    }

    public static func outcome(_ state: NativeBattleGameplayState) -> NativeConquestOutcome? {
        let player = countryStatus(state, owner: state.playerOwner)
        if player.defeated {
            return .defeat(reason: "country-extinction", owner: state.playerOwner, status: player)
        }
        let hostile = hostileOwners(state)
        if !hostile.isEmpty && hostile.allSatisfy({ countryStatus(state, owner: $0).defeated }) {
            return .victory(reason: "hostile-powers-extinct", owners: hostile)
        }
        return nil
    }
}
